include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,358 @@
+    lowerBound(lowerBound),
+    upperBound(upperBound),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type MOEAD::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+
+  // Weight vectors, where each one of them represents a decomposition.
+  std::vector<arma::vec> weights(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    weights[i] = arma::vec(numObjectives, arma::fill::randu);
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> weightNeighbourIndices(populationSize, neighbourhoodSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    // To temporarily store the distance between weights(i) and each other weights.
+    arma::vec distances(populationSize);
+    for (size_t j = 0; j < populationSize; j++)
+    {
+      distances(j) = 0;
+      for (size_t w = 0; w < numObjectives ; w++)
+      {
+        distances(j) += std::pow(weights[i](w) - weights[j](w), 2);
+      }
+      distances(j) = std::sqrt(distances(j));
+    }
+    arma::uvec sortedIndices = arma::stable_sort_index(distances, "descend");
+    for (size_t iter = 1; iter <= neighbourhoodSize; iter++)
+      weightNeighbourIndices(i, iter-1) = sortedIndices(iter);
+  }
+
+  // 1.3 Random generation of the initial population.
+  std::vector<MatType> population(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    population[i] = arma::randu<MatType>(iterate.n_rows, iterate.n_cols)
+        - 0.5 + iterate;
+
+  // 1.3 F-value initialisation for the population.
+  std::vector<arma::vec> FValue(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    FValue[i].resize(numObjectives);
+  EvaluateObjectives(population, objectives, FValue);
+
+  // 1.4 Initialize the ideal point z.
+  arma::vec idealPoint(numObjectives);
+  idealPoint.fill(-1/.0);
```

Comment:
Not sure I get the idea of using`-1/.0`  here, can you elaborate on that?

Reply:
Update of `idealPoint` happens as follows,

idealPoint(j) = max(idealPoint(j), F<sub>j</sub>(candidate)), where F<sub>j</sub> is the j<sup>th</sup> objective.

So initially all points in idealPoint are set to -INF, so after the first iteration, idealPoint will basically be holding the values of the first candidate we add to externalPopulation. The paper said to initialize idealPoint with a problem specific method. I was not sure what it meant so I went with this approach.

Reply:
So might be better to initialize with `std::numeric_limits<ElemType>::max()`?